Resolve OOM When Reading Large Logs in Webserver

jason810496 opened the pull request:
related issue: #45079 
related PR: #45129
related discussion on slack: https://apache-airflow.slack.com/archives/CCZRF2U5A/p1736767159693839

## Why

In short, this PR aims to eliminate OOM issues by:  
- Replacing full log sorting with a **K-Way Merge**  
- Making the entire log reading path **streamable** (using `yield` generators instead of returning a list of strings)  

More detailed reasoning is already described in the linked issue.

Due to too many conflicts with the old PR (#45129), this PR reworks the changes on top of the latest `FileTaskHandler`.

## What

This PR ports the original changes from #45129 to the current version of `FileTaskHandler` with the following updates:

- Fixed line-splitting error when reading in chunks using buffered line-splitting with a carry-over  
- Adopted the new log metadata structure  
- Introduced buffering for the log reader  

## Note: Recent Changes in `FileTaskHandler`

- Introduced `StructuredLogMessage` to represent each log record #46827
- Added `RemoteLogIO` interface for remote log handling #48491

jason810496 committed:
Add note for new usage of LogMetadata

jason810496 committed:
Add _stream_parsed_lines_by_chunk

jason810496 committed:
Refactor _read_from_local/logs_server as return stream

jason810496 committed:
Refactor _interleave_logs with K-Way Merge

jason810496 committed:
Add _get_compatible_log_stream

jason810496 committed:
Refactor _read method to return stream with compatible interface

- Add compatible interface for executor, remote_logs
- Refactor skip log_pos with skip for each log source

jason810496 committed:
Refactor log_reader to adapt stream

jason810496 committed:
Fix _read_from_local open closed file error

jason810496 committed:
Refactor LogReader by yielding in batch

jason810496 committed:
Add ndjson header to get_log openapi schema

jason810496 committed:
Fix _add_log_from_parsed_log_streams_to_heap
- Add comparator for StructuredLogMessage
- Refactor parsed_log_streams from list to dict for removing empty logs

jason810496 committed:
Fix _interleave_logs dedupe logic
- should check the current logs with default timestamp

jason810496 committed:
Refactor test_log_handlers
- Fix events utils
- Add convert_list_to_stream, mock_parsed_logs_factory utils
- Fix the following test after refactoring FileTaskHandler
    - test_file_task_handler_when_ti_value_is_invalid
    - test_file_task_handler
    - test_file_task_handler_running
    - test_file_task_handler_rotate_size_limit
    - test__read_when_local
    - test__read_served_logs_checked_when_done_and_no_local_or_remote_logs
    - test_interleave_interleaves
    - test_interleave_logs_correct_ordering
    - test_interleave_logs_correct_dedupe
- Add new test for refactoring FileTaskHandler
    - test__stream_lines_by_chunk
    - test__log_stream_to_parsed_log_stream
    - test__sort_key
    - test__is_sort_key_with_default_timestamp
    - test__is_logs_stream_like
    - test__add_log_from_parsed_log_streams_to_heap

jason810496 committed:
Move test_log_handlers utils to test_common

jason810496 committed:
Fix unit/celery/log_handlers test

jason810496 committed:
Fix mypy-providers static check

jason810496 committed:
Fix _get_compatible_log_stream
- sequential yield instead of parallel yield for all log_stream

jason810496 committed:
Fix amazon task_handler test

jason810496 committed:
Fix wask task handler test

jason810496 committed:
Fix elasticsearch task handler test

jason810496 committed:
Fix opensearch task handler test

jason810496 committed:
Fix TaskLogReader buffer
- don't concat buffer with empty str, yield directly from buffer

jason810496 committed:
Fix test_log_reader

jason810496 committed:
Fix CloudWatchRemoteLogIO.read mypy

jason810496 committed:
Fix test_gcs_task_handler

jason810496 committed:
Fix core_api test_log

jason810496 committed:
Fix CloudWatchRemoteLogIO._event_to_str dt format

jason810496 committed:
Fix TestCloudRemoteLogIO.test_log_message

jason810496 committed:
Fix es/os task_hander convert_list_to_stream

jason810496 committed:
Fix compact tests

jason810496 committed:
Refactor es,os task handler for 3.0 compact

jason810496 committed:
Fix compat for RedisTaskHandler

jason810496 committed:
Fix ruff format for test_cloudwatch_task_handler after rebase

jason810496 committed:
Fix 2.10 compat TestCloudwatchTaskHandler

jason810496 committed:
Fix 3.0 compat test for celery, wasb

Fix wasb test, spelling

jason810496 committed:
Fix 3.0 compat test for gcs

jason810496 committed:
Fix 3.0 compat test for cloudwatch, s3

uranusjr commented on airflow-core/src/airflow/api_fastapi/core_api/openapi/v1-rest-api-generated.yaml:
```diff
@@ -6383,7 +6383,7 @@ paths:
           - application/json
-          - text/plain
+          - application/x-ndjson
           - '*/*'
```
No idea about the implication, why is `text/plain` removed from explicitly listed?

jason810496 replied:
Since PR #46827 returns the `Content-Type` as `application/x-ndjson`, I’ve updated it to explicitly use `application/x-ndjson` to make it more consistent and appropriate.

https://github.com/apache/airflow/pull/46827/files#diff-18e153220119c0579d8b8030d13daaa04a623b38f9854ffff8702a2bbe68f4f8R123-R130

uranusjr commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/log.py:
```diff
@@ -133,24 +135,23 @@ def get_log(
+            continuation_token=encoded_token, content=list(structured_log_stream)
+        )
+
+    # text/plain, or something else we don't understand. Return raw log content in ndjson format with StreamingResponse
```
Is it a better idea to default to ndjson (and streaming), or a plain json response? I wonder if defaulting to a simple response is easier for simple cases.

jason810496 replied:
Actually, I was initially thinking of only providing a streaming `ndjson` response. However, for backward compatibility, I decided to retain the plain `json` format as well.

The last `list(structured_log_stream)` part breaks the memory-efficient streaming approach. The `json` serialization is resource-intensive and has, in some cases, caused the worker to be killed due to memory pressure. Because of this, I consider the `json` format more of a special-case fallback.

In my opinion, it's more appropriate for the frontend UI to adapt by specifying the corresponding `Accept` header format. ( That's why I opened issue #50333 to propose switching to the `ndjson` response format in the frontend )

uranusjr replied:
I’m not thinking about cases where the client asks for a format correctly, but those catch-all fallbacks. Our front _shouldn’t_ use those in any cases, but a custom client might (like when you want to debug the endpoint directly with postman or something). Defaulting to streaming on an unrecognised Accept header might make those cases more difficult, but I guess it’s not a big problem since you should just set the header correctly anyway. But in that case, why do we do this at all? Why not just return 406 Not Acceptable instead?

jason810496 replied:
Agreed, I see your point. Setting the default format to JSON will indeed make it easier for users to debug.  
I'll update the default format to JSON, thanks for bringing this up!

jason810496 replied:
Just resolved, thanks for the review.

jason810496 committed:
Set get_log API default response format to JSON

uranusjr commented on airflow-core/tests/unit/api_fastapi/core_api/routes/public/test_log.py:
Do we need some tests on the Accept logic?

jason810496 replied:
Sure, I will add later on.

uranusjr commented on providers/amazon/tests/unit/amazon/aws/log/test_cloudwatch_task_handler.py:
```diff
@@ -282,14 +282,27 @@ def test_read(self, monkeypatch):
+            if AIRFLOW_V_3_0:
+                assert metadata == {"end_of_log": False, "log_pos": 3}
+            else:
+                assert metadata == {"end_of_log": False, "first_time_read": False}
```
Hmm, we’ll need some backward compatibility measures. It’s not a good idea to just change the log format like this, especially since providers can have mismatch versions to Airflow Core.

jason810496 replied:
I see. I was thinking that all log sources should be able to resume reading from the last position, which is why I introduced the metadata change.

However, you're right. we should still consider backward compatibility for both providers and Airflow core. I'll remove the current metadata change and raise a separate PR focused on "enabling all log sources to resume from the last read position (instead of skipping in aggregated logs)" after this one is merged.

Mark PR as draft during refactoring.

jason810496 committed:
Remove "first_time_read" key in log metadata

jason810496 committed:
Remove "<source>_log_pos" key in log metadata

jason810496 committed:
Add LogStreamCounter for backward compatibility

jason810496 committed:
Remove "first_time_read" with backward "log_pos" for tests

- test_log_reader
- test_log_handlers
- test_cloudwatch_task_handler
- test_s3_task_handler
- celery test_log_handler
- test_gcs_task_handler
- test_wasb_task_handler
- fix redis_task_handler
- fix log_pos

jason810496 committed:
Fix RedisTaskHandler compatibility

jason810496 committed:
Fix chores in self review

- Fix typo in _read_from_logs_server
- Remove unused parameters in _stream_lines_by_chunk
- read_log_stream
    - Fix doc string by removing outdate note
    - Only add buffer for full_download
- Add test ndjson format for get_log API

jason810496 committed:
Fine-tune HEAP_DUMP_SIZE

uranusjr commented on providers/elasticsearch/src/airflow/providers/elasticsearch/version_compat.py:
```diff
@@ -33,3 +40,21 @@ def get_base_airflow_version_tuple() -> tuple[int, int, int]:
+        :param input_logs: List of structured log messages.
+        :return: A generator yielding structured log messages.
+        """
+        yield from input_logs
```
Isn’t this just `iter(input_logs)`? (for what we use it for)

This probably does not need to exist.

uranusjr replied:
And I suspect we don’t even need the call in many cases…

jason810496 replied:
Good question — yes, `iter(input_logs)` would work inline for simple cases, but in this context, I'm being strict with type annotations. The final goal is to make it `yield` from the first function that read the logs for the whole log reading path.

That’s why I’m using `Generator[...]` instead of `Iterable[...]`. It gives `mypy` a more precise view of what's going on — particularly that the function is a generator (and not,for example, a list).

So I add this util to make the type annotation correct, WDYT?
Or we should have `Iterable[...]` and replace this util with `iter`?

uranusjr replied:
I think Iterable is good enough. Mypy would recognise it can _either_ be a list or an iterator, and raise errors if you use the result in a way that an iterator can’t support (e.g. `len()`). I suspect it makes backward compatibility code easier to write too.

jason810496 replied:
Agree, I will make it an Iterable and this utils can be removed.

jason810496 replied:
Thanks @uranusjr for the suggestion, set the type annotation to `Iterable` is more clear.
I just replace the util with `iter` and removed the buffer for log reader.

jason810496 committed:
Replace get_compatible_output_log_stream with iter

jason810496 committed:
Remove buffer in log_reader

jason810496 committed:
Fix log_id not found compact for es_task_handler

pierrejeambrun reviewed (commented):
Nice, thanks for the benchmark, the memory improvement seems promising.

Do you know why it looks like the total response time is much longer after refactoring? Is that inherent to the stream solution? Also I'm not sure we care that much because the user will be able to see logs after the first chunck retrieval, wich might still be faster.

Lee-W commented on airflow-core/src/airflow/utils/log/log_stream_counter.py:
```diff
@@ -0,0 +1,151 @@
+        self,
+        stream: LogHandlerOutputStream,
+        threshold: int,
+    ):
```
```suggestion
    ) -> None:
```

Lee-W commented on airflow-core/src/airflow/utils/log/log_stream_counter.py:
```diff
@@ -0,0 +1,151 @@
+        self._disk_lines: int = 0
+        self._tmpfile = None
+
+    def _flush_buffer_to_disk(self):
```
```suggestion
    def _flush_buffer_to_disk(self) -> None:
```

Lee-W commented on airflow-core/src/airflow/utils/log/log_stream_counter.py:
```diff
@@ -0,0 +1,151 @@
+        self._tmpfile.flush()
+        self._buffer.clear()
+
+    def _capture(self):
```
```suggestion
    def _capture(self) -> None:
```

Lee-W commented on airflow-core/src/airflow/utils/log/log_stream_counter.py:
```diff
@@ -0,0 +1,151 @@
+
+                with open(self._tmpfile.name, encoding="utf-8") as f:
+                    for line in f:
+                        yield StructuredLogMessage.model_validate_json(line.strip())
```
```suggestion
                    yield from (StructuredLogMessage.model_validate_json(line.strip()) for line in f)
```

Lee-W commented on airflow-core/src/airflow/utils/log/log_stream_counter.py:
```diff
@@ -0,0 +1,151 @@
+            if len(self._buffer) >= self._threshold:
+                self._flush_buffer_to_disk()
+
+    def _cleanup(self):
```
```suggestion
    def _cleanup(self) -> None:
```

Lee-W commented on airflow-core/src/airflow/utils/log/log_stream_counter.py:
```diff
@@ -0,0 +1,151 @@
+            self._buffer.append(log)
+
+            if len(self._buffer) >= self._threshold:
+                self._flush_buffer_to_disk()
```
```suggestion
        while True:
            self._buffer = list(islice(self._stream, self._threshold))

            if self._buffer:
                self._flush_buffer_to_disk()
            else:
                break
```

not sure, but might be a little better 🤔

jason810496 replied:
Yes, we can use `itertools.isslice` to get up to `self._threshold` lines.
I just refactored with a small change of this function and adjusted the `structured_logs` fixture.

Lee-W commented on airflow-core/src/airflow/utils/log/log_stream_counter.py:
```diff
@@ -0,0 +1,151 @@
+    )
+
+
+class LogStreamCounter:
```
I don't quite understand this naming. This class seems to do more than just counting. Is this convention learned from somewhere else?

jason810496 replied:
The only way to get the total log line is to iterating through the whole log stream.

However, to make it memory efficient, we need to flush partial of logs to the temp file and also have the ability to replay the log stream after counting.

This is the only purpose of this class, and I come out with LogStreamCount as name. Maybe LogStreamLineCounter is more comprehensible?

Lee-W replied:
The counter is the part I'm confused about 🤔. What about  Accumulator? Collector?

jason810496 replied:
Accumulator is indeed more accurate in this scenario.

I will rename Counter with Accumulator, thanks for suggestion.

Lee-W commented on airflow-core/src/airflow/utils/log/log_reader.py:
```diff
@@ -83,20 +86,15 @@ def read_log_stream(self, ti: TaskInstance, try_number: int | None, metadata: di
-                yield (log if isinstance(log, str) else log.model_dump_json()) + "\n"
+            log_stream, out_metadata = self.read_log_chunks(ti, try_number, metadata)
+            for log in log_stream:
+                yield log.model_dump_json() + "\n"
```
```suggestion
            yield from (f"{log.model_dump_json()}\n" for log in log_stream)
```

Lee-W reviewed (commented):
I'll take several rounds of review to finish this one. Left a few early suggestions

jason810496 committed:
Fix review comments
- rename LogStreamCounter as LogStreamAccumulator
- simply for-yield with yield-from in log_reader
- add type annotation for LogStreamAccumulator

jason810496 committed:
Refactor LogStreamAccumulator._capture method
- use itertools.isslice to get chunk

jason810496 committed:
Merge branch 'main' into refactor/rework-webserver-oom-for-large-log-read

jason810496 committed:
Fix type hint, joinedload for ti.dag_run after merge

Lee-W commented on airflow-core/src/airflow/utils/log/file_task_handler.py:
```diff
@@ -36,24 +39,66 @@
+"""
+RawLogStream: TypeAlias = Generator[str, None, None]
+"""Raw log stream, containing unparsed log lines."""
+LegacyLogResponse: TypeAlias = tuple[LogSourceInfo, LogMessages]
```
By `Legacy` do you mean airflow 2 or before this PR

jason810496 replied:
I mean "before this PR" , including Airflow 2 and Airflow 3 (3.0.1, 3.02) for backward compatibility.

Lee-W replied:
yep, maybe worth mentioning it in the comments

Lee-W commented on airflow-core/src/airflow/utils/log/file_task_handler.py:
```diff
@@ -166,17 +242,133 @@ def _parse_log_lines(
 
-def _interleave_logs(*logs: str | LogMessages) -> Iterable[StructuredLogMessage]:
-    min_date = pendulum.datetime(2000, 1, 1)
+def _sort_key(timestamp: datetime | None, line_num: int) -> int:
```
```suggestion
def _create_sort_key(timestamp: datetime | None, line_num: int) -> int:
```

not sure whether it's better 🤔 it kinda confused me when I saw `sort_key` is a none and `_sort_key` looks like a verb that returns `sort_key`

jason810496 replied:
Agree, adding the verb for prefix is more readable

Lee-W commented on airflow-core/src/airflow/utils/log/file_task_handler.py:
```diff
@@ -166,17 +242,133 @@ def _parse_log_lines(
+        sort_key, line = heapq.heappop(heap)
+        if line != last or _is_sort_key_with_default_timestamp(sort_key):  # dedupe
+            yield line
+        last = line
```
```suggestion
        sort_key, line = heapq.heappop(heap)
        if line != last or _is_sort_key_with_default_timestamp(sort_key):  # dedupe
            yield line
        last = line
```

nit: we might be able to extract it as a function as it looks similar to the lines above

Lee-W commented on airflow-core/src/airflow/utils/log/file_task_handler.py:
```diff
@@ -166,17 +242,133 @@ def _parse_log_lines(
+    del parsed_log_streams
+
+
+def _is_logs_stream_like(log):
```
```suggestion
def _is_logs_stream_like(log) -> bool:
```

Lee-W commented on airflow-core/src/airflow/utils/log/file_task_handler.py:
```diff
@@ -166,17 +242,133 @@ def _parse_log_lines(
+
+def _is_logs_stream_like(log):
+    """Check if the logs are stream-like."""
+    return isinstance(log, chain) or isinstance(log, GeneratorType)
```
```suggestion
    return isinstance(log, (chain, GeneratorType))
```

Lee-W commented on airflow-core/src/airflow/utils/log/file_task_handler.py:
```diff
@@ -166,17 +242,133 @@ def _parse_log_lines(
+    log_streams: list[RawLogStream] = []
+    for log_message in log_messages:
+        # Append the log stream to the list
+        log_streams.append(_stream_lines_by_chunk(io.StringIO(log_message)))
```
```suggestion
    log_streams: list[RawLogStream] = [
        _stream_lines_by_chunk(io.StringIO(log_message))
        for log_message in log_messages
    ]
```

Lee-W commented on airflow-core/src/airflow/utils/log/file_task_handler.py:
```diff
@@ -36,24 +39,66 @@
+"""The legacy format of log messages, represented as a single string blob to be parsed later."""
 LogSourceInfo: TypeAlias = list[str]
 """Information _about_ the log fetching process for display to a user"""
 LogMetadata: TypeAlias = dict[str, Any]
```
not sure whehter TypeDict would be better

jason810496 replied:
The `LogMessages` is for both released Airflow 2(`list[str]` and Airflow 3(`ist["StructuredLogMessage"]`) compatible type annotation, so I prefer to keep it.

Lee-W replied:
I mean `LogMetadata`

jason810496 replied:
I changed all the `dict[str,Any]` type annotation for `LogMetadata` with `TypeDict`.

Thanks for the suggestion, this is indeed more clear. 
Though we need to add `# type: ignore [assignment]` for someplace for mypy when assign the dict to the variable when handling the case that `metadata` is empty.

Lee-W commented on airflow-core/src/airflow/utils/log/file_task_handler.py:
```diff
@@ -370,22 +562,38 @@ def _read(
-            sources, remote_logs = self._read_remote_logs(ti, try_number, metadata)
-
+            sources, logs = self._read_remote_logs(ti, try_number, metadata)
+            if not len(logs):
```
```suggestion
            if not logs:
```

Lee-W commented on airflow-core/src/airflow/utils/log/log_reader.py:
```diff
@@ -72,7 +73,7 @@ def read_log_chunks(
         end.
         """
-        return self.log_handler.read(ti, try_number, metadata=metadata)
+        return self.log_handler.read(ti, try_number, metadata)
```
why this change?

jason810496 replied:
It was my typo, I reverted my change for this line.

Lee-W commented on airflow-core/src/airflow/utils/log/log_stream_accumulator.py:
```diff
@@ -0,0 +1,154 @@
+            if len(self._buffer) >= self._threshold:
+                self._flush_buffer_to_disk()
+            else:  # If there are no more lines to capture, exit the loop.
+                break
```
```suggestion
            # If there are no more lines to capture, exit the loop.
            if len(self._buffer) < self._threshold:
                break
            self._flush_buffer_to_disk()
```

jason810496 replied:
I would like to keep my change.

IMO, for small logs (log lines small then `HEAP_DUMP_SIZE`), we should avoid flushing the logs to disk to make it faster to response.
Additionally, they're small enough, so buffer them in memory will be fine.

Lee-W replied:
hmmm... I'm a bit confused here. I thought what I suggested is basically the same thing, but in a different order. is there anything I missed 🤔

jason810496 replied:
Sorry, I overlooked indent and mis-understood the logic. You are right, we means the same thing and I adopt your change, thanks!

Lee-W replied:
it seems to be missed here

Lee-W commented on airflow-core/src/airflow/utils/log/log_stream_accumulator.py:
```diff
@@ -0,0 +1,154 @@
+            os.remove(self._tmpfile.name)
+            self._tmpfile = None
+
+    def get_total_lines(self) -> int:
```
not sure whether making it a property makes sense.

Lee-W commented on airflow-core/src/airflow/utils/log/log_stream_accumulator.py:
```diff
@@ -0,0 +1,154 @@
+        """
+        return self._disk_lines + len(self._buffer)
+
+    def get_stream(self) -> StructuredLogStream:
```
not sure whether making it a property makes sense.

jason810496 replied:
Sure, making them as properties is more appropriate. 
I refactored the `get_total_lines` and `get_stream` to class property.

Lee-W commented on airflow-core/src/airflow/utils/serve_logs.py:
```diff
@@ -135,7 +135,7 @@ def validate_pre_signed_url():
     @flask_app.route("/log/<path:filename>")
     def serve_logs_view(filename):
-        return send_from_directory(log_directory, filename, mimetype="application/json", as_attachment=False)
+        return send_from_directory(log_directory, filename, mimetype="text/plain", as_attachment=False)
```
may i know why did we change this?

jason810496 replied:
The log files is formatted in NDJSON currently ( for Airflow 2, they are even not store in JSON-like format). So changing the response `mimetype` just to make the the response type more accurate, will not effect on actual logic.

Additionally, in this PR, we will parsed the log server response in stream  with `log_streams.append(_stream_lines_by_chunk(io.TextIOWrapper(response.raw)))`, so setting the mimetype to `text/plain` is also more suitable.

jason810496 committed:
Replace _sort_key as _create_sort_key

jason810496 committed:
Add _flush_logs_out_of_heap common util

jason810496 committed:
Fix review nits

 - _is_logs_stream_like
    - add type annotation
    - reduce to 1 isinstance call
- construct log_streams in _get_compatible_log_stream inline
- use TypeDict for LogMetadata
- remove len(logs) to check empty
- revert typo of self.log_handler.read in log_reader
- log_stream_accumulator
    - refactor flush logic
    - make totoal_lines as property
    - make stream as property

jason810496 committed:
Merge branch 'main' into refactor/rework-webserver-oom-for-large-log-read

jason810496 committed:
Fix mypy errors after merge

jason810496 committed:
Fix redis task handler test

jason810496 committed:
Merge branch 'main' into refactor/rework-webserver-oom-for-large-log-read

Lee-W commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/log.py:
```diff
@@ -138,24 +144,24 @@ def get_log(
-        logs, metadata = task_log_reader.read_log_chunks(ti, try_number, metadata)
-        encoded_token = None
+    if accept == Mimetype.NDJSON:  # only specified application/x-ndjson will return streaming response
+        log_stream = task_log_reader.read_log_stream(ti, try_number, metadata)  # type: ignore[arg-type]
```
Let's add a comment on why `type: ignore[arg-type]` is needed

jason810496 replied:
Agree. I added comments for all LogMetadata related mypy ignore.

Lee-W commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/log.py:
```diff
@@ -138,24 +144,24 @@ def get_log(
+
+    # application/json, or something else we don't understand.
+    # Return JSON format, which will be more easily for users to debug.
+    structured_log_stream, out_metadata = task_log_reader.read_log_chunks(ti, try_number, metadata)  # type: ignore[arg-type]
```
same here

Lee-W commented on airflow-core/src/airflow/api_fastapi/core_api/routes/public/log.py:
```diff
@@ -138,24 +144,24 @@ def get_log(
+    # application/json, or something else we don't understand.
+    # Return JSON format, which will be more easily for users to debug.
+    structured_log_stream, out_metadata = task_log_reader.read_log_chunks(ti, try_number, metadata)  # type: ignore[arg-type]
+    encoded_token = None
```
I feel like we can extract the token logic here as a function, but that can be done in a following up PR

jason810496 replied:
Sure. I just add this comment to my backlog. I will raise follow-up PR after this get merged.

Lee-W commented on airflow-core/src/airflow/utils/log/file_task_handler.py:
```diff
@@ -19,51 +19,111 @@
-LogMessages: TypeAlias = Union[list["StructuredLogMessage"], list[str]]
-"""The log messages themselves, either in already sturcutured form, or a single string blob to be parsed later"""
+LogMessages: TypeAlias = list[str]
+"""The legacy format of log messages, represented as a single string blob to be parsed later."""
```
Let's improve the comment by explaining what this `legacy` means. If I'm not misunderstood, it's pre-3.0.2

jason810496 replied:
Yes, the legacy LogMessages will include `3.0.2`, so I refined the description for it.

Lee-W commented on airflow-core/src/airflow/utils/log/file_task_handler.py:
```diff
@@ -19,51 +19,111 @@
+    end_of_log: bool
+    log_pos: NotRequired[int]
+    # the following attributes are used for Elasticsearch and OpenSearch log handlers
+    offset: NotRequired[str | int]
```
It's interesting that it could be a `str` in some cases 🤔  but I guess it's more for backward compatibiity

jason810496 replied:
This is a special one related to `OpensearchTaskHandler` and `ElasticsearchTaskHandler`, which is not related to backward compatibiity, it's because `Large offset numbers will get JSON.parsed incorrectly on the client. Sending as a string prevents this issue.`

I had copied the comment back as well.

Lee-W commented on airflow-core/src/airflow/utils/log/file_task_handler.py:
```diff
@@ -166,17 +248,147 @@ def _parse_log_lines(
+    :param heap: heap to store log records
+    :param parsed_log_streams: dict of parsed log streams
+    """
+    log_stream_to_remove: list[int] | None = None
+    for idx, log_stream in parsed_log_streams.items():
+        record: ParsedLog | None = next(log_stream, None)
+        if record is None:
+            if log_stream_to_remove is None:
+                log_stream_to_remove = []
+            log_stream_to_remove.append(idx)
+            continue
+        # add type hint to avoid mypy error
+        record = cast("ParsedLog", record)
+        timestamp, line_num, line = record
+        # take int as sort key to avoid overhead of memory usage
+        heapq.heappush(heap, (_create_sort_key(timestamp, line_num), line))
+    # remove empty log stream from the dict
+    if log_stream_to_remove is not None:
+        for idx in log_stream_to_remove:
+            del parsed_log_streams[idx]
```
```suggestion
    log_stream_to_remove: list[int] []
    for idx, log_stream in parsed_log_streams.items():
        record: ParsedLog | None = next(log_stream, None)
        if record is None:
            log_stream_to_remove.append(idx)
            continue
        # add type hint to avoid mypy error
        record = cast("ParsedLog", record)
        timestamp, line_num, line = record
        # take int as sort key to avoid overhead of memory usage
        heapq.heappush(heap, (_create_sort_key(timestamp, line_num), line))

    # remove empty log stream from the dict
    for idx in log_stream_to_remove:
        del parsed_log_streams[idx]
```

not sure whether I understand it correctly, but I'm not sure whether we need the additional logic here

jason810496 replied:
Yes, your understanding is correct. However, I intend to make the `log_stream_to_remove` list to be initialized lazily. Since for most of the time, we don't have any log steam need to be add to the `log_stream_to_remove` list, we are just adding front of each log streams to the heap.

IMO, initializing the `log_stream_to_remove` lazily can reduce the overhead. Though I'm not sure whether initializing a list for each call will be large or not?  I will remain current implementation, no strong opinion on this.

Lee-W replied:
If that's the case, we probably should doc it as well. I bet there will be someone who wants to "refactor" it if we don't doc here. I also did some quick experiments, and yes there could be some overhead we'd better avoid when

Lee-W commented on airflow-core/src/airflow/utils/log/file_task_handler.py:
```diff
@@ -166,17 +248,147 @@ def _parse_log_lines(
+    ]
+
+    for log_stream in log_streams:
+        yield from log_stream
```
```suggestion
    yield from chain.from_iterable(
        _stream_lines_by_chunk(io.StringIO(log_message)) for log_message in log_messages
   )
```

we can make it a standalone variable if we want to keep the type

jason810496 replied:
Adapt with `chain.from_iterable`, first time to use this utility!

Lee-W commented on airflow-core/src/airflow/utils/log/file_task_handler.py:
```diff
@@ -423,12 +649,23 @@ def _read(
+            # skip log stream until the last position
+            if metadata and "log_pos" in metadata:
+                for _ in range(metadata["log_pos"]):
+                    next(out_stream, None)
```
not sure we can use `islice` here

jason810496 replied:
Yes, we can use `islice` here as well.

Lee-W commented on airflow-core/src/airflow/utils/log/log_reader.py:
```diff
@@ -91,24 +92,19 @@ def read_log_stream(
-            # Update the metadata dict in place so caller can get new values/end-of-log etc.
-
-            for log in logs:
-                # It's a bit wasteful here to parse the JSON then dump it back again.
-                # Optimize this so in stream mode we can just pass logs right through, or even better add
-                # support to 307 redirect to a signed URL etc.
```
Is this comment still needed?

jason810496 replied:
I think we don't need anymore.

Lee-W replied:
sounds good. just want to confirm it

Lee-W commented on airflow-core/tests/unit/utils/log/test_stream_accumulator.py:
```diff
@@ -0,0 +1,163 @@
+                    timestamp=LOG_START_DATETIME.add(seconds=i),
+                    level="INFO",
+                    message=f"Test log message {i + 1}",
+                )
```
```suggestion
            yield from (
                StructuredLogMessage(
                    event=f"test_event_{i + 1}",
                    timestamp=LOG_START_DATETIME.add(seconds=i),
                    level="INFO",
                    message=f"Test log message {i + 1}",
                )
                for i in range(LOG_COUNT)
            )
```

Lee-W replied:
Hmmmm... we're using `LOG_COUNT = 20` above. Should it be an argument instead?

jason810496 replied:
I fix the nit for `yield from`.

> Hmmmm... we're using LOG_COUNT = 20 above. Should it be an argument instead?

However, for current test cases in this module, the expected result is depends on `LOG_COUNT = 20` so I tend to not change make it an argument.

Lee-W replied:
Ah... I misread it. Please ignore the "Hmmm..." comment.

Lee-W commented on providers/redis/tests/unit/redis/log/test_redis_task_handler.py:
```diff
@@ -111,7 +112,12 @@ def test_read(self, ti):
 
         if AIRFLOW_V_3_0_PLUS:
-            assert logs == (["Line 1\nLine 2"], {"end_of_log": True})
+            if get_base_airflow_version_tuple() <= (3, 0, 2):
```
```suggestion
            if get_base_airflow_version_tuple() < (3, 0, 3):
```

Lee-W commented on providers/opensearch/tests/unit/opensearch/log/test_os_task_handler.py:
```diff
@@ -204,6 +204,7 @@ def test_read(self, ti):
             "on 2023-07-09 07:47:32+00:00"
         )
         if AIRFLOW_V_3_0_PLUS:
+            logs = list(logs)
```
why do we need this?

jason810496 replied:
After this refactor, the return type of `.read` method will be iterator, so we need to add the `list(logs)` to make the validation in test easier.

Lee-W replied:
got it. makes sense

Lee-W commented on providers/opensearch/src/airflow/providers/opensearch/log/os_task_handler.py:
```diff
@@ -346,7 +347,9 @@ def _read(
         """
         if not metadata:
-            metadata = {"offset": 0}
+            metadata = {"offset": 0}  # type: ignore[assignment]
```
Let's add a comment on why this ignoring is needed

Lee-W commented on devel-common/src/tests_common/test_utils/file_task_handler.py:
```diff
@@ -0,0 +1,77 @@
+    from airflow.utils.log.file_task_handler import ParsedLog, StructuredLogMessage
+
+
+def events(logs: Iterable[StructuredLogMessage], skip_source_info=True) -> list[str]:
```
```suggestion
def extract_events(logs: Iterable[StructuredLogMessage], skip_source_info=True) -> list[str]:
```

Lee-W commented on airflow-core/tests/unit/utils/test_log_handlers.py:
```diff
@@ -188,14 +189,16 @@ def task_callable(ti):
-        assert any(re.search(target_re, e) for e in events(log)), "Logs were " + str(log)
+        assert any(re.search(target_re, e) for e in events(log_handler_output_stream)), "Logs were " + str(
+            log_handler_output_stream
+        )
```
```suggestion
        assert any(re.search(target_re, e) for e in events(log_handler_output_stream)), f"Logs were {log_handler_output_stream}"
```

Lee-W commented on airflow-core/tests/unit/utils/test_log_handlers.py:
```diff
@@ -724,34 +775,244 @@ def test_parse_timestamps():
 
+def test__sort_key():
+    # assert _sort_key should return int
+    assert isinstance(_create_sort_key(pendulum.parse("2022-11-16T00:05:54.278000-08:00"), 10), int)
```
shouldn't we just use the expected int to ensure the value is correct?

Lee-W commented on airflow-core/tests/unit/utils/test_log_handlers.py:
```diff
@@ -724,34 +775,244 @@ def test_parse_timestamps():
+    assert _is_logs_stream_like(log_stream) == expected
+
+
+def test__add_log_from_parsed_log_streams_to_heap():
```
Let's add some comments to clarify what we're checking here. I'm not getting the length checking part

jason810496 replied:
Sure, I had added detail for this test to describe the key point to validate in each steps.

Lee-W replied:
Thanks a lot!

Lee-W commented on devel-common/src/tests_common/test_utils/file_task_handler.py:
```diff
@@ -0,0 +1,77 @@
+                i,
+                structured_log,
+            )
+        )
```
```suggestion
    parsed_logs: list[ParsedLog] = [
        (
            start_datetime + pendulum.duration(seconds=i),
            i,
            StructuredLogMessage(timestamp=None, event=f"{event_prefix} Event {i}")
        )
        for i in range(count):
    ]    
```

I also notice the timestamp is set to `None` here. Is it expected

jason810496 replied:
This is a real typo mistake. Nice catch!

jason810496 committed:
Refactor _capture logic in LogStreamAccumulator

jason810496 committed:
Add comments for ingore LogMetadata TypeDict

jason810496 committed:
Add comment for offset; Fix commet for LogMessages

jason810496 committed:
Refactor with from_iterable, islice

jason810496 committed:
Fix nits in test

- refactor structured_logs fixtures in TestLogStreamAccumulator
- use f-strign in test_file_task_handler
- assert actual value of _create_sort_key
- add details comments in test__add_log_from_parsed_log_streams_to_heap

jason810496 committed:
Refactor test_utils

jason810496 committed:
Add comment for lazy initialization

jason810496 committed:
Fix error handling for _stream_lines_by_chunk

jason810496 committed:
Merge branch 'main' into refactor/rework-webserver-oom-for-large-log-read

jason810496 committed:
Fix mypy error after merge

potiuk commented on providers/celery/tests/unit/celery/log_handlers/test_log_handlers.py:
```diff
@@ -78,14 +81,24 @@ def test__read_for_celery_executor_fallbacks_to_worker(self, create_task_instanc
             fth._read_from_logs_server = mock.Mock()
-            fth._read_from_logs_server.return_value = ["this message"], ["this\nlog\ncontent"]
+
+            # compact with 2.x and 3.x
```
```suggestion
            # compat with 2.x and 3.x
```

Super-nit

Lee-W commented on providers/redis/tests/unit/redis/log/test_redis_task_handler.py:
```diff
@@ -111,7 +112,12 @@ def test_read(self, ti):
 
         if AIRFLOW_V_3_0_PLUS:
-            assert logs == (["Line 1\nLine 2"], {"end_of_log": True})
+            if get_base_airflow_version_tuple() < (3, 0, 3):
```
This probably need to be taken care.

Lee-W commented on devel-common/src/tests_common/test_utils/file_task_handler.py:
```diff
@@ -0,0 +1,76 @@
+    logs = iter(logs)
+    if skip_source_info:
+
+        def is_source_group(log: StructuredLogMessage):
```
```suggestion
        def is_source_group(log: StructuredLogMessage) -> bool:
```

Lee-W reviewed (approved):
Completed a quick round, but the `3, 0, 3` issue needs addressing. Everything else looks good.

jason810496 committed:
Fix final review nits

jason810496 committed:
Merge branch 'main' into refactor/rework-webserver-oom-for-large-log-read

jason810496 committed:
Fix mypy error

jason810496 committed:
Merge branch 'main' into refactor/rework-webserver-oom-for-large-log-read

ashb commented on providers/amazon/src/airflow/providers/amazon/aws/log/cloudwatch_task_handler.py:
```diff
@@ -206,8 +197,6 @@ def get_cloudwatch_logs(self, stream_name: str, task_instance: RuntimeTI):
             log_stream_name=stream_name,
             end_time=end_time,
         )
-        if AIRFLOW_V_3_0_PLUS:
-            return list(self._event_to_dict(e) for e in events)
```
@jason810496 This line means that logs from Cloudwatch are now showing as raw-json strings rather then rendered nicely. What was the thinking here?

jason810496 replied:
> What was the thinking here?


After rethink, the removal of compact here is wrong.

For the other remote handler, the change should be like following:
```diff
                if AIRFLOW_V_3_0_PLUS:
-                    return missing_log_message, metadata
+                   from airflow.utils.log.file_task_handler import StructuredLogMessage
+
+                   return [StructuredLogMessage(event=missing_log_message)], metadata
```

`missing_log_message` is `str`, and this PR was merged as `3.0.3 plus`
- For `3.0.0-3.0.2`, the return type of logs is `str`
- For `3.0.3 plus` the return type of logs is `list[StructuredLogMessage]`

The refactored FileTaskHandler does handle both cases.

I think this mistake removal is because AWS cloudwatch handler is the first remote handler I work on and there are some changes in the development iteration then I forget to update the logic of cloudwatch handler after the changes.

ashb replied:
@jason810496 A sort of fix (but not a complete one) that I knocked up https://github.com/astronomer/airflow/commit/afbf6f514be23b671225ccc4d275a76d77e04989

Could you finish it off and get logs for cloudwatch working properly please.

I was using https://www.localstack.cloud/ (the free version) and I set these env vars in breeze to test it

```
AWS_ENDPOINT_URL=http://localstack-main:4566
AWS_ACCESS_KEY_ID="test" AWS_SECRET_ACCESS_KEY="test"
AWS_DEFAULT_REGION="us-east-1"
AIRFLOW__LOGGING__REMOTE_BASE_LOG_FOLDER=cloudwatch://arn:aws:logs:us-east-1:000000000000:log-group:test
AIRFLOW__LOGGING__REMOTE_LOGGING=True
```

(And then create the log group via the localstack gui, or using the AWS cli)

We also need to test latest provider against 3.0.2 and main airflow with released Amazon provider version.

ashb commented on providers/amazon/src/airflow/providers/amazon/aws/log/cloudwatch_task_handler.py:
```diff
@@ -222,7 +211,8 @@ def _event_to_dict(self, event: dict) -> dict:
         event_dt = datetime.fromtimestamp(event["timestamp"] / 1000.0, tz=timezone.utc)
-        formatted_event_dt = event_dt.strftime("%Y-%m-%d %H:%M:%S,%f")[:-3]
+        # Format a datetime object to a string in Zulu time without milliseconds.
+        formatted_event_dt = event_dt.strftime("%Y-%m-%dT%H:%M:%SZ")
```
Ditto: why did you change this? This seems to be unrealted to OOM behavoviour, and should have been in a different PR if it was going to be changed.

Please keep changes focused to a single thing.

jason810496 replied:
Sorry, this is my mistake, I thought the behavior with the change should be same but it turn out to be the root cause of why the cloudwatch handler is not working properly.

ashb commented on airflow-core/src/airflow/utils/log/file_task_handler.py:
```diff
@@ -19,51 +19,114 @@
-LogMessages: TypeAlias = list["StructuredLogMessage"] | list[str]
-"""The log messages themselves, either in already sturcutured form, or a single string blob to be parsed later"""
+LogMessages: TypeAlias = list[str]
+"""The legacy format of log messages before 3.0.2"""
```
@jason810496 This is before 3.0.3 now right? I think we didn't include this in 3.0.3 in the end

jason810496 replied:
Yes, this patch will be included in `3.0.4` release.
